### PR TITLE
Fix an issue for multi CUPTI sessions

### PR DESCRIPTION
**Pull request by kaixih**
When using an external profiler tools (like NVIDIA Nsight Systems), some collected components might be missing in the profiling results if the TF profiler is used in the python script. This PR fixes this issue by explicitly checking if the CUPTI tracer is successfully enabled or not. If not, we return an error msg instead of a OK status.

FYI. @nluehr

### CODE REVIEW

File: tensorflow/core/profiler/internal/gpu/cupti_tracer.h [hunk 1]
```diff
@@ -87,7 +87,7 @@ class CuptiTracer {
   bool IsAvailable() const;
   bool NeedRootAccess() const { return need_root_access_; }
 
-  void Enable(const CuptiTracerOptions& option, CuptiTraceCollector* collector);
+  bool Enable(const CuptiTracerOptions& option, CuptiTraceCollector* collector);
```
Comment (sanjoy) [inline tensorflow/core/profiler/internal/gpu/cupti_tracer.h:90]: Please document what the return value means.

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Review by sanjoy — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by jbaiocchi** [comment]
It is insufficient to check that CUPTI is already enabled, as this means the TensorFlow Profiler is active and incurring other overheads (e.g., TraceMe annotations). The correct thing to do when using another profiling tool is to fully disable the TF Profiler. Typically TF Profiler is off by default. One exception is when using the Keras TensorBoard callback [1]: the profile_batch argument should be set to 0 but the default is 2. Unfortunately changing the default would break backwards compatibility. One alternative is to ensure that models using the callback offer a flag as in [2]. Note that TF itself is a library that should not define command-line flags. An acceptable alternative would be an environment variable that disables the callback and is set by the script that enables Nsight.

[1] https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/TensorBoard
[2] https://github.com/tensorflow/models/blob/2986bcafb9eaa8fed4d78f17a04c4c5afc8f6691/official/vision/image_classification/resnet/common.py

**Comment by yisitu** [comment]
Shouldn't the user disable device tracing in TF profiler before running the application under another profiler (Nsight Systems)? Is this not possible for some cases?

**Comment by jbaiocchi** [comment]
An alternative solution is provided in:
https://github.com/tensorflow/tensorflow/commit/6d7e583c890e852319629ee00bf8ca87c233475b

You can set the environment variable TF_DISABLE_PROFILING to 1 to disable the profiler. It should prevent CUPTI (and all other profiler sources) from being activated by the internal TF profiler when an external profiler is active.

**Comment by kaixih** [comment]
Hi @jbaiocchi , thanks for adding the env variable. I just tested it but it seems it works differently than I thought.

Say, now I have a script using the TF profiler. If I set TF_DISABLE_PROFILING=1 and run it, the execution will error out with `tensorflow.python.framework.errors_impl.UnavailableError: Another profiler session is active.` even when I haven't use other profilers yet.

But I thought what we wanted is by setting the TF_DISABLE_PROFILING, the internal TF profiler will be skipped silently (or maybe with some warning). Then, the users can use any external profilers.

For example, the script demo.py uses TF profiler. And this is what we wanted to achieve:
1. `python demo.py` # run with internal profiler
2. `TF_DISABLE_PROFILING=1 python demo.py` # still run but without any profiling (Unfortunately it will error out now) 
3. `TF_DISABLE_PROFILING=1 nsys profile python demo.py` # run with external nsys profiler
4. `nsys profile python demo.py` # undefined for now.

**Comment by jbaiocchi** [comment]
Setting TF_DISABLE_PROFILING=1 is still the way to go.
The problem (and root cause of this bug) is that the Keras TensorBoard callback enables profiling by default.
The profiler API raises exceptions that the callback was not catching. I have a fix under review.

**Comment by jbaiocchi** [comment]
Please verify if the following commit is sufficient or there's still some issue:
https://github.com/tensorflow/tensorflow/commit/53ff2339b3c05a6efbe5bdbdd6c40b8973fb8e17

**Comment by kaixih** [comment]
Just tried with the nightly, but it seems the error is still on.
Below is the script that I used:
```python
import tensorflow as tf

tf.profiler.experimental.start('logdir_path')

inputs = tf.ones((10, 100, 100, 4))
conv = tf.keras.layers.Conv2D(
    10, (3, 3), strides=(1, 1), padding='valid',)
outputs = conv(inputs)
_ = outputs.numpy()

tf.profiler.experimental.stop()

```
Then, I ran it with:
```
root@96002906402c:/home/kernel_tests/cupti_tracer# TF_DISABLE_PROFILING=1 python -u cupti_tracer.py
2021-02-27 00:46:22.652281: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-02-27 00:46:23.534106: W tensorflow/core/profiler/lib/profiler_lock.cc:39] TensorFlow Profiler is permanently disabled by env var TF_DISABLE_PROFILING.
2021-02-27 00:46:23.534211: I tensorflow/core/profiler/lib/profiler_session.cc:158] Profiler session tear down.
Traceback (most recent call last):
  File "cupti_tracer.py", line 3, in <module>
    tf.profiler.experimental.start('logdir_path')
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/profiler/profiler_v2.py", line 122, in start
    _profiler.start(logdir, opts)
tensorflow.python.framework.errors_impl.UnavailableError: Another profiler session is active.

```

**Comment by jbaiocchi** [comment]
The behavior you are seeing is intended. tf.profiler.experimental.start and tf.profiler.experimental.stop are documented to raise exceptions. Setting TF_DISABLE_PROFILING=1 makes then raise. If this was python user code, it would be the responsibility of the user to catch exceptions. The commit addresses the bug in the TensorboardCallback:  it was making calls to the profiler without catching exceptions.

**Comment by kaixih** [comment]
The above code is not from users but myself to quick test the cupti profiler.

> The commit addresses the bug in the TensorboardCallback: it was making calls to the profiler without catching exceptions.

So, do you mean if the script uses the Tensorboard, the profiler will be silently skipped when we set TF_DISABLE_PROFILING?

For example, I just tried the following tensorboard_tf2.py in the nightly build:
```python
import tensorflow as tf

tensorboard_dir = 'logdir_path'
training_hooks = []
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=tensorboard_dir)
training_hooks.append(tensorboard_callback)

model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(8))
model.add(tf.keras.layers.Dense(1))
model.compile(optimizer='sgd', loss='mse')


x = tf.ones((100, 4))
y = tf.ones((100,))

# This builds the model for the first time:
model.fit(x, y, batch_size=32, epochs=10, callbacks=training_hooks)
```
However, I still got the same error:
```
root@b6c450ec0f0c:/home/kernel_tests/cupti_tracer# TF_DISABLE_PROFILING=1 python -u tensorboard_tf2.py
2021-03-01 18:44:06.635002: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-03-01 18:44:07.528315: W tensorflow/core/profiler/lib/profiler_lock.cc:39] TensorFlow Profiler is permanently disabled by env var TF_DISABLE_PROFILING.
2021-03-01 18:44:07.528436: I tensorflow/core/profiler/lib/profiler_session.cc:158] Profiler session tear down.
Traceback (most recent call last):
  File "tensorboard_tf2.py", line 5, in <module>
    tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=tensorboard_dir)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/callbacks.py", line 2139, in __init__
    self._init_profile_batch(profile_batch)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/callbacks.py", line 2348, in _init_profile_batch
    self._start_profiler(logdir='')
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/callbacks.py", line 2527, in _start_profiler
    profiler.start(logdir=logdir)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/profiler/profiler_v2.py", line 122, in start
    _profiler.start(logdir, opts)
tensorflow.python.framework.errors_impl.UnavailableError: Another profiler session is active.
```

**Comment by jbaiocchi** [comment]
We had a mismatch between the python exception and the C++ error code. This is now fixed:
https://github.com/tensorflow/tensorflow/commit/f07914d9d7e6eaf97b913fed985303da93544a7d
Please give it a try. Thanks!

**Comment by kaixih** [comment]
Thx. Just verified it with the nightly.

One more thing to confirm: so it seems we need to cherry-pick these three commits for this `TF_DISABLE_PROFILING` feature, right? @jbaiocchi 
https://github.com/tensorflow/tensorflow/commit/6d7e583c890e852319629ee00bf8ca87c233475b
https://github.com/tensorflow/tensorflow/commit/53ff2339b3c05a6efbe5bdbdd6c40b8973fb8e17
https://github.com/tensorflow/tensorflow/commit/f07914d9d7e6eaf97b913fed985303da93544a7d

**Comment by jbaiocchi** [comment]
Yes, those are the 3 required commits for TF_DISABLE_PROFILING.

**Comment by kaixih** [comment]
Thx. Based on https://github.com/tensorflow/tensorflow/pull/46544#issuecomment-790067558, close this PR.

**Pull request closed without merging**